### PR TITLE
[MM-24507]: sticky "reply to thread..." input

### PR DESCRIPTION
**Pull request by michelengelen**
#### Summary
the reply to thread input was always scrolling out of view for very long threads. This PR adds the sticky attribute to the input, so that it stays in view all the time.

#### Ticket Link
[MM-24507](https://mattermost.atlassian.net/browse/MM-24507)

#### Release Note
```release-note
made the reply to thread input sticky
```

### CODE REVIEW

**Comment by koox00** [comment]
I think that wouldn't work though for large threads, more than 100 or 200 I cannot remember the exact number, 
but the list is virtualized and so is the `CreateComment`, which means it won't be rendered if scroll up a large thread list...

**Comment by michelengelen** [comment]
> I think that wouldn't work though for large threads, between 100 and 200 I cannot remember the exact number, but the list is virtualized and so is the `CreateComment`, which means it won't be rendered if scroll up a large thread list...

Fair point ... Question: is there a case where the `CreateComment` component is NOT rendered?

**Comment by michelengelen** [comment]
> We could try to move it out the list, then the only problem to solve would be sticking it to the bottom of a really small list.

Well .... that proves to be quite difficult due to a somewhat messy CSS and DOM structure ... I am gonna give it a shot though!

**Comment by koox00** [comment]
> Fair point ... Question: is there a case where the `CreateComment` component is NOT rendered?

You mean product wise, no, well the wrapper that `ThreadViewer` is showing is always rendered, 
but the actual text input is not rendered always, 
see https://github.com/mattermost/mattermost-webapp/blob/master/components/threading/virtualized_thread_viewer/create_comment.tsx#L45-L71

> Well .... that proves to be quite difficult due to a somewhat messy CSS and DOM structure ... I am gonna give it a shot though!

Yep I know. I had tried that before, let me see if I have a branch on that.

**Comment by michelengelen** [comment]
> > Fair point ... Question: is there a case where the `CreateComment` component is NOT rendered?
> 
> You mean product wise, no, well the wrapper that `ThreadViewer` is showing is always rendered, but the actual text input is not rendered always, see https://github.com/mattermost/mattermost-webapp/blob/master/components/threading/virtualized_thread_viewer/create_comment.tsx#L45-L71
> 

Well, that leaves the option of rendering it to the actual component, so from that perspective it should be fine to extract that from the virtualised List

**Comment by koox00** [comment]
I had tried one other thing, but I can't remember results, making the `dynamic-sized-list` support `position: sticky` items.

**Comment by michelengelen** [comment]
> I had tried one other thing, but I can't remember results, making the `dynamic-sized-list` support `position: sticky` items.

I am just going to give extraction a shot next week .. thanks for checking in here! :D

**Comment by koox00** [comment]
thanks for taking up this issue!

**Review by nevyangelova — APPROVED** [review]

**Comment by matthewbirtch** [comment]
This seems to be working great - I can't seem to find any cases where things break. At some point we really need to clean up the reply input area. There is a lot of wasted space under it and I've always thought it odd that we have a reply button here unlike the main channel area. However, that probably needs to be designed and scoped fully with a separate PR. Nice work on this!

**Review by matthewbirtch — APPROVED** [review]

**Comment by esethna** [comment]
Approving but let's please make sure to test the cases that we ran into previously that caused us to revert this a few times in the past:

- I would suggest QA reads through the comment history on the ticket to understand the challenges faces with this previously: https://mattermost.atlassian.net/browse/MM-24507?focusedCommentId=100585

Please also test this in its entirely on all supported browsers

**Review by esethna — APPROVED** [review]

**Comment by michelengelen** [comment]
ok, I found the issue @koox00 mentioned myself ... as we are using a virtualized list for displaying the thread, the comment field does not get rendered when we reach a certain threshold (61+ replies) and scroll to the top. I have a screenshot of this:
Screenshot 2022-01-10 at 10 10 35

**Comment by koox00** [comment]
@michelengelen yep, sorry if I wasn't clear before, but that's what I meant :)
we'll need to either extract the input from the list, or make `DynamicSizeList` support sticky components.

**Review by koox00 — CHANGES_REQUESTED** [review]
just to block merging

**Comment by michelengelen** [comment]
Thanks! ... I know that making the dynamic list accept sticky elements is too much of a change (and unneeded) ... I am currently extracting the comment field  from the list and try to add it manually! :D

**Comment by michelengelen** [comment]
this should now be working ... although the server does not seem to accept replies atm (it happened on another PR lately too) ... does anyone might know whats happening there?

**Comment by michelengelen** [comment]
@koox00, @matthewbirtch, @esethna and @nevyangelova could you please re-review the changes? THX!

**Comment by matthewbirtch** [comment]
@michelengelen I just tested replying to a thread and it's failing every time. It doesn't seem to be a connection issue because I can post a new root post - just not a reply post in the RHS

image

**Comment by michelengelen** [comment]
> @michelengelen I just tested replying to a thread and it's failing every time. It doesn't seem to be a connection issue because I can post a new root post - just not a reply post in the RHS

That's strange since I did. It change the way replies get submitted. I will have a look regardless

**Comment by matthewbirtch** [comment]
> > @michelengelen I just tested replying to a thread and it's failing every time. It doesn't seem to be a connection issue because I can post a new root post - just not a reply post in the RHS
> 
> That's strange since I did. It change the way replies get submitted. I will have a look regardless

It may be an issue with the cloud servers....see the comment from @jgilliam17 here: https://github.com/mattermost/mattermost-webapp/pull/8882#issuecomment-1011733245

File: sass/components/_post-right.scss [hunk 1]
```diff
@@ -137,7 +137,17 @@
...
+    .post-right__content--inner-wrapper {
+        position: absolute;
+        display: flex;
+        overflow: hidden;
+        flex-direction: column;
+        inset: 0;
+    }
+
     .post-create__container {
+        position: sticky !important;
```
Comment (nevyangelova) [inline sass/components/_post-right.scss:149]: I know this is an ongoing issue with webapp and thats why I would like to start this campaign to remove all !importants and find way to optimise our CSS, so I would try to make this selector more specific and not use important here;.

**Review by nevyangelova — CHANGES_REQUESTED** [review]

**Review by esethna — CHANGES_REQUESTED** [review]
I'm seeing an issue where it's not pinning me to the bottom as I add more replies (

**Review by koox00 — APPROVED** [review]

**Comment by michelengelen** [comment]
@nevyangelova and @esethna could you please re-review? Thanks! 🙇🏼

**Review by nevyangelova — APPROVED** [review]

**Review by esethna — APPROVED** [review]
Looks good to me, but I will stress again that this should be thoroughly tested on all browsers keeping in mind the things that have caused this to be reverted in the past: https://github.com/mattermost/mattermost-webapp/pull/9482#issuecomment-1007687737 cc// @jgilliam17

**Comment by jgilliam17** [comment]
@matthewbirtch here's an invite to test server that can be used for testing.  https://pr9482.test.mattermost.cloud/signup_user_complete/?id=k78nj3ga5tdqjfamq8xxutxhne
Since it's built via cloud bot using latest commit from this PR it will need to be manually upgraded each time changes are made.

**Comment by matthewbirtch** [comment]
> @matthewbirtch here's an invite to test server that can be used for testing. https://pr9482.test.mattermost.cloud/signup_user_complete/?id=k78nj3ga5tdqjfamq8xxutxhne Since it's built via cloud bot using latest commit from this PR it will need to be manually upgraded each time changes are made.

Thanks @jgilliam17!

**Comment by matthewbirtch** [comment]
@michelengelen a few things I just noticed:
1. When I open a long thread in the RHS, it doesn't quite open to the bottom of the thread. 
image

2. When CRT is on and new replies come in from another user while I'm scrolled up in the thread, a 'New Replies' banner is supposed to display anchored to the bottom, but I'm not seeing that.

**Comment by esethna** [comment]
@jgilliam17 @michelengelen how's this PR coming along?

**Comment by michelengelen** [comment]
@esethna I prioritized the growth-spike. Will pick this back up once that is done.

**Comment by esethna** [comment]
@michelengelen please ping Jelena when this is ready for a QA re-review :)

**Comment by michelengelen** [comment]
@esethna and @jgilliam17 ready for another round.

@matthewbirtch fixed both issues you found with the RHS thread viewer! 🙇🏼

**Review by matthewbirtch — APPROVED** [review]
Awesome work @michelengelen!

**Comment by jgilliam17** [comment]
Thanks @michelengelen Looks great - tested on macOS, Win 10 and Ubuntu,  various browsers and desktop app

few small issues
1. In mobile view, [user] is typing... pushes the texbox up and in turn covers the last post. This is not an issue on non-mobile as `user is typing...` text is within textbox links area (help, preview etc.) 

https://user-images.githubusercontent.com/52937121/156224076-07e86a33-1a89-4aab-b6db-e85953659d2a.mov

2. When reacting to last message using `cmd shift \` the emoji is not immediately visible, has to be scrolled in.

3. Small issue with responsiveness when changing  browser size.
Steps:
- start with about half screen width browser, open rhs and scroll to the bottom of the long thread
- double-click on header to fullscreen the browser 
Observed: thread scrolls when changing size, no longer seeing last post

4. When opening a thread in the rhs, where last reply is long, but can still be fully displayed on the rhs, the thread opens sightly scrolled up from the bottom 

https://user-images.githubusercontent.com/52937121/156240812-229d9b3b-5cf1-4942-b962-1cd6cce8d184.mov

5. Not reproducing consistently: 
On thread view, when I am viewing an older post (scrolled away from the bottom) and I dismiss "New Replies" toast in order to remain in the same scrolled position - l would get scrolled to the bottom when additional new messages come in. 
Again, just wanted to mention here that I have seen this, but not able to repro, I suspect there must be something else at play not just new message arriving. While testing I was logged in, multiple sessions and devices at the same time so that could be one of the reasons.

**Comment by michelengelen** [comment]
@jgilliam17 thanks for the findings ... commenting below your points! ⬇️ 

>     1. In mobile view, [user] is typing... pushes the texbox up and in turn covers the last post. This is not an issue on non-mobile as `user is typing...` text is within textbox links area (help, preview etc.)

I did not know about that. For some reason we are hiding the `Help` link in mobile view (probably because we do the same in the channels view). I am currently discussing the possibility of removing the `Reply` button so we have a more uniform way of showing textinputs. I will fix once we come to a conclusion there, ok?
 
>     2. When reacting to last message using `cmd shift \` the emoji is not immediately visible, has to be scrolled in.

That is a behavior I cannot test. since some hotkeys do not work on a `QWERTZ` keyboard layout. (IDK why, will ask in the web team)
 
>     3. Small issue with responsiveness when changing  browser size.

should now be fixed!

>     4. When opening a thread in the rhs, where last reply is long, but can still be fully displayed on the rhs, the thread opens sightly scrolled up from the bottom

I cannot reproduce that. Did you try with a specific viewport size?

>     5. Not reproducing consistently:
>        On thread view, when I am viewing an older post (scrolled away from the bottom) and I dismiss "New Replies" toast in order to remain in the same scrolled position - l would get scrolled to the bottom when additional new messages come in.
>        Again, just wanted to mention here that I have seen this, but not able to repro, I suspect there must be something else at play not just new message arriving. While testing I was logged in, multiple sessions and devices at the same time so that could be one of the reasons.

I cannot reproduce this one as well.

**Comment by jgilliam17** [comment]
Thanks @michelengelen 
1. Sounds good, we can wait for decision 
3. I am still seeing this on the test server, is it possible changes were not applied?
4. This is reproducing for me in all window sizes on my 15 inch macbook, except in full screen.

**Comment by jgilliam17** [comment]
Thanks @michelengelen 
I'll re-run e2es, removal of Reply button on the RHS is going to fail few tests - this will have to wait for master to be fixed, we're currently not able to start a cycle.
- Something I missed last time around - uploaded files push textbox up, obscuring most recent messages on the rhs
Screen Shot 2022-03-07 at 12 04 11 PM

**Comment by michelengelen** [comment]
@jgilliam17 the findings you pointed out here (the reply box obscuring the content by expanding) are standard browser behavior ... I don't know if we want to mess with that since all scrolling adjustments are relatively hefty, not to speak of the event and height listeners we need to add. This might be worth discussing with UX.

WDYT @matthewbirtch ?

**Comment by matthewbirtch** [comment]
> @jgilliam17 the findings you pointed out here (the reply box obscuring the content by expanding) are standard browser behavior ... I don't know if we want to mess with that since all scrolling adjustments are relatively hefty, not to speak of the event and height listeners we need to add. This might be worth discussing with UX.
> 
> WDYT @matthewbirtch ?

The behaviour on the thread should be the same as in the channel. If I'm scrolled to the bottom of the channel and I add an attachment, the input area expands and the channel remains scrolled to the bottom. We should do the same with the thread view.

**Comment by michelengelen** [comment]
> > @jgilliam17 the findings you pointed out here (the reply box obscuring the content by expanding) are standard browser behavior ... I don't know if we want to mess with that since all scrolling adjustments are relatively hefty, not to speak of the event and height listeners we need to add. This might be worth discussing with UX.
> > WDYT @matthewbirtch ?
> 
> The behaviour on the thread should be the same as in the channel. If I'm scrolled to the bottom of the channel and I add an attachment, the input area expands and the channel remains scrolled to the bottom. We should do the same with the thread view.

Hey @matthewbirtch ... After a quick investigation this is not as easy as it sounds ... Since the overall layout of the main channel view is different (the textinput is tuck to the bottom all the time) I would need to extend the functionality of the ThreadView to allow for re-calculating scrolling position when something changes on the attachments.

I am generally not against something like this, but manually scrolling is always a pretty hefty operation (performance is pretty bad for that), so we have mainly 3 options here:
1. either remain with the current behavior, or ...
2. add new functionality to scroll when changes appear, or ...
3. refactor the Threadviewer to a different layout (I would close this PR in that case)

**Comment by matthewbirtch** [comment]
> Hey @matthewbirtch ... After a quick investigation this is not as easy as it sounds ... Since the overall layout of the main channel view is different (the textinput is tuck to the bottom all the time) I would need to extend the functionality of the ThreadView to allow for re-calculating scrolling position when something changes on the attachments.
> 
> I am generally not against something like this, but manually scrolling is always a pretty hefty operation (performance is pretty bad for that), so we have mainly 3 options here:
> 
> 1. either remain with the current behavior, or ...
> 2. add new functionality to scroll when changes appear, or ...
> 3. refactor the Threadviewer to a different layout (I would close this PR in that case)

If we can do #2 without significant perf issues, I'd prefer we go that route.

One other thing I just noticed:
- The send button should only show on mobile widths. It's showing on a desktop width for me and is causing overlapping with the message text
image

**Comment by michelengelen** [comment]
@matthewbirtch I did take a look at how we could handle the height change from the fileupload and without a refactor this is not easily done. I would opt for merging it as is (after fixing the failed tests) and follow this up with a new PR. Would that be OK with you?

**Comment by michelengelen** [comment]
@matthewbirtch I updated the mobile view (removed the send-button from non-mobile view) and merged the current master to fix merge conflicts. Could you have a last look and check if everything is in place now? Thanks!

**Comment by matthewbirtch** [comment]
Looks like the test server didn't get updated @michelengelen

**Review by matthewbirtch — APPROVED** [review]
Seems good to go @michelengelen

**Comment by michelengelen** [comment]
@jgilliam17 this is now ready for another round of QA review

**Comment by jgilliam17** [comment]
Thanks @michelengelen 
Few [E2E tests](https://mattermost-cypress-report.s3.amazonaws.com/185883-832126b-onprem-ent-webapp-pr-9482-mm-ee-test:832126b/mochawesome.html) related to RHS failed 
I'll update and re-run e2es

**Comment by jgilliam17** [comment]
There are still some tests failing to find Add comment/reply button. Can you please take a look. Thanks. 🙂  
Latest [E2E report ](https://mattermost-cypress-report.s3.amazonaws.com/185919-2189cca-onprem-ent-webapp-pr-9482-mm-ee-test:2189cca/mochawesome.html)

**Comment by michelengelen** [comment]
hey @jgilliam17 the tests should now be fixed... I will re-run the e2e tests now

**Comment by jgilliam17** [comment]
@michelengelen Here is the link to latest [E2E report ](https://mattermost-cypress-report.s3.amazonaws.com/188545-0346eee-onprem-ent-webapp-pr-9482-mattermost/mm-ee-test0346eee/mochawesome.html)

**Comment by michelengelen** [comment]
@jgilliam17 I fixed the failing tests (kinda):
1. MM-T209: got fixed
2. MM-T3307: I removed this one since we do not have the reply button anymore. There is a different problem with the previews and I already posted a question about that in the UX-Channel [here](https://community-daily.mattermost.com/core/pl/hoefmk1xetnn7reeuf854njepc).
3. MM-T3318: this git fixed by merging master. Apparently the gif link was broken.

I will run e2e tests again after the checks are green.

**Comment by jgilliam17** [comment]
Thanks @michelengelen and sorry about all the back and forth.
Tested, looks good to merge.
[E2E report ](https://mattermost-cypress-report.s3.amazonaws.com/191171-e32090f-onprem-ent-webapp-pr-9482-mattermost/mm-ee-teste32090f/mochawesome.html)looks good, no PR related failures.

**Review by jgilliam17 — APPROVED** [review]

**Pull request merged**